application/comit_node/src/comit_client/bam.rs

```diff
@@ -150,69 +151,141 @@ fn to_json_value(
+                        info!("Connection to {} established", socket_addr);
+                        let final_connection_state = Arc::clone(&new_connection_state);
+                        let mut new_connection_state = new_connection_state.lock().unwrap();
+
+                        let senders: Vec<_> = match *new_connection_state {
+                            ConnectionState::Connecting(ref mut senders) => senders.drain(..).collect(),
+                            ref invalid_state => {
+                                error!(
+                                    "Somehow a client we transitioned to {:?} while we were connecting",
+                                    invalid_state
+                                );
+                                vec![]
+                            }
+                        };
+
+                        match socket {
+                            Ok(socket) => {
+                                let (client, connection) = json_connection_from_socket(socket, Config::default());
+                                let client = Arc::new(BamClient::new(socket_addr, client));
+                                for sender in senders {
+                                    let _ = sender.send(Ok(client.clone()));
+                                }
+                                *new_connection_state = ConnectionState::Connected(client);
+
+                                Box::new(connection.then(move |result| {
+                                    {
+                                        let mut final_connection_state =
+                                            final_connection_state.lock().unwrap();
+                                        *final_connection_state = ConnectionState::Disconnected;
+                                    }
+
+                                    match result {
+                                        Ok(_) => {
+                                            info!("Connection to {} was closed", socket_addr);
+                                        }
+                                        Err(e) => {
+                                            error!(
+                                                "Connection to {} prematurely closed: {:?}",
+                                                socket_addr, e
+                                            );
+                                        }
+                                    }
+                                    Ok(())
+                                }))
+                            }
+
+                            Err(e) => {
+                                error!("Error connecting to {}: {:?}", socket_addr, e);
+                                for sender in senders {
+                                    let _ = sender.send(Err(ClientFactoryError::Connection(e.kind())));
+                                }
+                                *new_connection_state = ConnectionState::Disconnected;
```

Comment:
:thinking: Should we have a connection state `Failed`?

Reply:
Probably. I suggest we implement it when we want it for something though.